jupyterhub_traefik_proxy/proxy.py

```diff
@@ -19,11 +19,129 @@
+                    auth_username=self.traefik_api_username,
+                    auth_password=self.traefik_api_password,
+                )
+                resp_frontends = await AsyncHTTPClient().fetch(
+                    self.traefik_api_url + "/api/providers/" + provider + "/frontends",
+                    auth_username=self.traefik_api_username,
+                    auth_password=self.traefik_api_password,
+                )
+                backends_data = json.loads(resp_backends.body)
+                frontends_data = json.loads(resp_frontends.body)
+
+                if resp_backends.code == 200 and resp_frontends.code == 200:
+                    ready = (
+                        expected_backend in backends_data
+                        and expected_frontend in frontends_data
+                    )
+            except Exception as e:
+                backends_rc, frontends_rc = e.response.code
+                ready = False
+            finally:
+                return ready
+
+        await exponential_backoff(
+            _check_traefik_dynamic_conf_ready,
+            "Traefik route for %s configuration not available" % target,
+            timeout=20,
+        )
+
+    async def _wait_for_static_config(self, provider):
+        async def _check_traefik_static_conf_ready():
+            """ Check if traefik loaded its static configuration from the
+            etcd cluster """
+            try:
+                resp = await AsyncHTTPClient().fetch(
+                    self.traefik_api_url + "/api/providers/" + provider,
+                    auth_username=self.traefik_api_username,
+                    auth_password=self.traefik_api_password,
+                )
+                rc = resp.code
+            except ConnectionRefusedError:
+                rc = None
+            except Exception as e:
+                rc = e.response.code
+            finally:
+                return rc == 200
+
+        await exponential_backoff(
+            _check_traefik_static_conf_ready,
+            "Traefik static configuration not available",
+            timeout=20,
+        )
+
+    def _stop_traefik(self):
+        self.log.info("Cleaning up proxy[%i]...", self.traefik_process.pid)
+        self.traefik_process.kill()
+        self.traefik_process.wait()
+
+    def _launch_traefik(self, config_type):
+        if config_type == "toml":
+            config_file_path = abspath(join(dirname(__file__), "traefik.toml"))
+            self.traefik_process = Popen(
+                ["traefik", "-c", config_file_path], stdout=None
+            )
+        elif config_type == "etcd":
+            self.traefik_process = Popen(
+                ["traefik", "--etcd", "--etcd.useapiv3=true"], stdout=None
+            )
+        else:
+            self.log.error(
+                "Configuration mode not supported \n. The proxy can only be configured through toml and etcd"
+            )
+            raise
```

Comment:
`raise` only works if there's an Exception. Try: `raise ValueError(...)` with the above message.